pages/workspace/workspace.html.js

```diff
@@ -288,11 +312,28 @@ function Workspace({ navTitle, showLogs, resLocals }) {
 
             socket.emit('startWorkspace', { workspace_id: workspaceId });
 
-            setInterval(() => {
+            const intervalId = setInterval(() => {
               socket.emit('heartbeat', { workspace_id: workspaceId }, (msg) => {
                 console.log('heartbeat, msg =', msg);
               });
             }, heartbeatInterval * 1000);
+
+            let timeoutId = null;
+            document.addEventListener('visibilitychange', () => {
+              if (document.visibilityState == 'hidden') {
+                timeoutId = setTimeout(() => {
+                  clearInterval(intervalId);
```

Comment:
Currently, we rely on disabling the heartbeat to shut down the container. Maybe we should use an explicit message to stop it? That would let us track a metric like `workspaces.stopped.visibility.count` if we were interested to know how many workspaces we're shutting down like this.

Reply:
I like this idea. The current behavior is due to the fact that we were assuming we wouldn't be able to detect why we went offline, but in this new situation it would be great to get more metrics. We could also register with [`beforeunload`](https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event) and do a similar thing then?

Reply:
Nice, I'll make this change and add a metric.

Not sure I understand what we'd do in an unload handler, can you elaborate?

Reply:
I was thinking that we could register a handler with `beforeunload()` that would also send a websocket message to stop the workspace. This would trigger in the case that the user closes the tab or navigates away. But maybe these actions will already trigger a `visibilityState` change?

HOWEVER, thinking about this whole idea further I now think that we should _not_ send any explicit "stop workspace" events and instead we should only rely on heartbeat timeouts. This is because of group workspaces where we have multiple students connected to the same backend workspace (e.g., the collaborative Jupyter workspace). We don't want to shut it down just because one student goes away because the others might still be working.

Reply:
I guess we could send the visibility timeout message so long as the server doesn't blindly shut the workspace down, but instead uses this as a trigger to check for recent heartbeats (basically running the usual workspace-timeout-check cron job, but just for this workspace). Maybe this is worth it for the metrics? But it's a fair bit of additional code and complexity that we'll be keeping around.

Reply:
Ahhhh, great point about group workspaces! That makes a lot of sense; I'll remove the explicit stop.

Reply:
Given this, it probably also makes sense to *not* immediately show a "stopped due to inactivity" message and instead rely on receiving a `change:state` message if when it ultimately stops. What do you think? I think in that case we'd also want to restart the heartbeat if the tab becomes visible again.

Reply:
Yes! So you are saying that the client will just do the following two things?

1. Send heartbeats whenever it's visible.
2. Respond to server messages like `change:state` by displaying the message and changing which frame is visible.

This seems very clean!

Reply:
That's what I'm thinking, yeah!